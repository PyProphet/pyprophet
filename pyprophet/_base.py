from dataclasses import dataclass, field
import os
import copy
import sys

from loguru import logger
from .io.util import (
    is_tsv_file,
    is_sqlite_file,
    is_parquet_file,
    is_valid_single_split_parquet_dir,
    is_valid_multi_split_parquet_dir,
)


@dataclass
class BaseIOConfig:
    """
    Base configuration class for I/O-related metadata and common attributes used across algorithms.

    Attributes:
        infile (str): Path to the input file (e.g., .osw, .parquet).
        outfile (str): Path to the output file to be written.
        file_type (str): Type of the input file (e.g., 'osw', 'parquet', 'parquet_split', 'tsv').
        subsample_ratio (float): Subsampling ratio for large data.
        level (str): Scoring level (e.g., 'ms1', 'ms2', 'transition', 'alignment').
        context (str): Context or mode in which the reader/writer operates
                       (e.g., 'score_learn', 'score_apply', 'ipf', 'level_context').
        prefix (str): Automatically derived from outfile (e.g., 'results/output' for 'results/output.osw').
    """

    infile: str
    outfile: str
    file_type: str = field(init=False)
    subsample_ratio: float
    level: str
    context: str
    prefix: str = field(init=False)

    def __post_init__(self):
        """
        Initialize the file_type and prefix attributes based on the input file.
        """

        infile = self.infile

        if is_sqlite_file(infile):
            if infile.endswith(".osw"):
                self.file_type = "osw"
            elif infile.endswith(".sqmass") or infile.endswith(".sqMass"):
                self.file_type = "sqmass"
        elif is_parquet_file(infile):
            self.file_type = "parquet"
        elif is_valid_single_split_parquet_dir(infile):
            self.file_type = "parquet_split"
        elif is_valid_multi_split_parquet_dir(infile):
            self.file_type = "parquet_split_multi"
        elif is_tsv_file(infile):
            self.file_type = "tsv"
        else:
            logger.critical(
                f"Failed to infer file type for: {infile}. Supported formats are: "
                ".osw, .sqmass/.sqMass, .parquet, split parquet directories (.oswpq/oswpqd), or .tsv files.\n"
                f"  - is_sqlite_file: {is_sqlite_file(infile)}\n"
                f"  - endswith .osw: {infile.endswith('.osw')}\n"
                f"  - endswith .sqmass/.sqMass: {infile.endswith('.sqmass') or infile.endswith('.sqMass')}\n"
                f"  - is_parquet_file: {is_parquet_file(infile)}\n"
                f"  - is_valid_single_split_parquet_dir: {is_valid_single_split_parquet_dir(infile)}\n"
                f"  - is_valid_multi_split_parquet_dir: {is_valid_multi_split_parquet_dir(infile)}\n"
                f"  - is_tsv_file: {is_tsv_file(infile)}"
            )

            sys.exit(1)

        self.prefix = os.path.splitext(
            self.outfile
        )[
            0
        ]  # TODO: use pathlib instead to avoid potential cases where the outfile is a directory (split_parquet) file.oswpq/, which would result the prefix in being file.oswpq instead of file

    def __str__(self):
        return (
            f"BaseIOConfig(\ninfile='{self.infile}'\noutfile='{self.outfile}'\n"
            f"file_type='{self.file_type}'\nsubsample_ratio={self.subsample_ratio}\n"
            f"level='{self.level}'\ncontext='{self.context}'\nprefix='{self.prefix}')"
        )

    def __repr__(self):
        return (
            f"BaseIOConfig(infile='{self.infile}', outfile='{self.outfile}', "
            f"subsample_ratio={self.subsample_ratio}, level='{self.level}', "
            f"context='{self.context}')"
        )

    def copy(self):
        """
        Return a deep copy of the config object.
        """
        return copy.deepcopy(self)
