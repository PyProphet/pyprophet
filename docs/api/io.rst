IO: Reading and Writing Data
=====================================================



.. automodule:: pyprophet.io
   :members:
   :inherited-members:

Abstract Base Classes
----------------------

.. currentmodule:: pyprophet.io

.. autosummary::
   :nosignatures:
   :toctree: generated/
   :template: class.rst

   _base.BaseReader
   _base.BaseWriter
   _base.BaseOSWReader
   _base.BaseOSWWriter
   _base.BaseParquetReader
   _base.BaseParquetWriter
   _base.BaseSplitParquetReader
   _base.BaseSplitParquetWriter
   dispatcher.ReaderDispatcher
   dispatcher.WriterDispatcher

These submodules provide specific implementations for reading and writing data for specific algorithms.

Scoring
----------------------

.. automodule:: pyprophet.io.scoring
   :no-members:
   :no-inherited-members:

.. currentmodule:: pyprophet.io.scoring

.. autosummary:: 
   :nosignatures:
   :toctree: generated/
   :template: class.rst

   osw.OSWReader
   osw.OSWWriter
   parquet.ParquetReader
   parquet.ParquetWriter
   split_parquet.SplitParquetReader
   split_parquet.SplitParquetWriter
   tsv.TSVReader
   tsv.TSVWriter

IPF
----------------------

.. automodule:: pyprophet.io.ipf
   :no-members:
   :no-inherited-members:

.. currentmodule:: pyprophet.io.ipf

.. autosummary:: 
   :nosignatures:
   :toctree: generated/
   :template: class.rst

   osw.OSWReader
   osw.OSWWriter
   parquet.ParquetReader
   parquet.ParquetWriter
   split_parquet.SplitParquetReader
   split_parquet.SplitParquetWriter

Levels Context
----------------------

.. automodule:: pyprophet.io.levels_context
   :no-members:
   :no-inherited-members:

.. currentmodule:: pyprophet.io.levels_context

.. autosummary:: 
   :nosignatures:
   :toctree: generated/
   :template: class.rst

   osw.OSWReader
   osw.OSWWriter
   parquet.ParquetReader
   parquet.ParquetWriter
   split_parquet.SplitParquetReader
   split_parquet.SplitParquetWriter

Export
----------------------

.. automodule:: pyprophet.io.export
   :no-members:
   :no-inherited-members:

.. currentmodule:: pyprophet.io.export

.. autosummary:: 
   :nosignatures:
   :toctree: generated/
   :template: class.rst

   osw.OSWReader
   osw.OSWWriter
   parquet.ParquetReader
   parquet.ParquetWriter
   split_parquet.SplitParquetReader
   split_parquet.SplitParquetWriter