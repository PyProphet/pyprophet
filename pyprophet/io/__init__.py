"""
The `io` module provides tools and utilities for handling input and output operations
in PyProphet. It supports various file formats, including SQLite (OSW),
Parquet, Split Parquet, and TSV, and provides functionality for reading, writing,
and validating data.

Submodules:
-----------
- `util`: Contains utility functions for file validation, schema inspection, and logging.
- `dispatcher`: Provides dispatcher classes for routing I/O configurations to the appropriate
  reader and writer implementations based on file type and context.
- `_base`: Defines abstract base classes and utility methods for implementing custom readers
  and writers for different data formats.

Dependencies:
-------------
- `pandas`
- `pyarrow`
- `duckdb`
- `sqlite3`
- `loguru`
- `click`

"""
