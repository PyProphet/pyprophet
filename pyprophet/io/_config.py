from dataclasses import dataclass, field
from typing import Any, Dict
import os
from ._base import BaseIOConfig


@dataclass
class ErrorEstimationConfig:
    # Global FDR & pi0
    parametric: bool
    pfdr: bool
    pi0_lambda: Any  # Could be float or List[float]
    pi0_method: str
    pi0_smooth_df: int
    pi0_smooth_log_pi0: bool

    # Local FDR
    lfdr_truncate: bool
    lfdr_monotone: bool
    lfdr_transformation: str
    lfdr_adj: bool
    lfdr_eps: float


@dataclass
class RunnerConfig:
    # Scoring / classifier options
    classifier: str
    ss_main_score: str
    main_score_selection_report: bool

    # XGBoost-related hyperparameters
    xgb_hyperparams: bool
    xgb_params: dict
    xgb_params_space: dict

    # Cross-validation settings
    xeval_fraction: float
    xeval_num_iter: int

    # Semi-supervised settings
    ss_initial_fdr: float
    ss_iteration_fdr: float
    ss_num_iter: int
    ss_score_filter: bool
    ss_use_dynamic_main_score: bool = field(init=False)

    # Grouping & statistical options
    group_id: str
    error_estimation_config: ErrorEstimationConfig

    # IPF options
    ipf_max_peakgroup_rank: int
    ipf_max_peakgroup_pep: float
    ipf_max_transition_isotope_overlap: float
    ipf_min_transition_sn: float

    # Glyco options
    glyco: bool
    density_estimator: str
    grid_size: int

    # Miscellaneous
    add_alignment_features: bool
    tric_chromprob: bool
    threads: int
    test: bool
    color_palette: str

    def __post_init__(self):
        # Check for auto main score selection
        if self.ss_main_score == "auto":
            # Set starting default main score
            self.ss_main_score = "var_xcorr_shape"
            self.use_dynamic_main_score = True
        else:
            self.use_dynamic_main_score = False


@dataclass
class RunnerIOConfig(BaseIOConfig):
    runner: RunnerConfig
    extra_writes: dict = field(init=False)

    def _post_init_(self):
        self.extra_writes = dict(self._extra_writes())

    def to_kwargs(self) -> Dict[str, Any]:
        return {
            "infile": self.infile,
            "outfile": self.outfile,
            "level": self.level,
            "prefix": self.prefix,
            **vars(self.runner),
        }

    def _extra_writes(self):
        """
        Generates paths for various output files based on the prefix provided.

        Yields:
            Tuple[str, str]: A tuple containing the name of the output file type and the corresponding file path.
        """
        yield "output_path", os.path.join(self.prefix + "_scored.tsv")
        yield "summ_stat_path", os.path.join(self.prefix + "_summary_stat.csv")
        yield "full_stat_path", os.path.join(self.prefix + "_full_stat.csv")
        yield "trained_weights_path", os.path.join(self.prefix + "_weights.csv")
        yield "trained_model_path_ms1", os.path.join(self.prefix + "_ms1_model.bin")
        yield "trained_model_path_ms1ms2", os.path.join(
            self.prefix + "_ms1ms2_model.bin"
        )
        yield "trained_model_path_ms2", os.path.join(self.prefix + "_ms2_model.bin")
        yield "trained_model_path_transition", os.path.join(
            self.prefix + "_transition_model.bin"
        )
        yield "report_path", os.path.join(self.prefix + "_report.pdf")


@dataclass
class IPFIOConfig(BaseIOConfig):
    ipf_filter_decoys: bool
    min_ipf_confidence: float


@dataclass
class LevelContextIOConfig(BaseIOConfig):
    context_level: str
    aggregation_method: str
