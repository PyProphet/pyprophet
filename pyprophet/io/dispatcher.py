"""
This module provides dispatcher classes for routing I/O configurations to the appropriate
reader and writer implementations.

The dispatchers determine the correct implementation based on the file type and context
(e.g., scoring, IPF, or levels context). Supported file types include:
- OSW
- Parquet
- Split Parquet
- TSV (for scoring only)

Classes:
- ReaderDispatcher: Routes configurations to the appropriate reader implementation.
- WriterDispatcher: Routes configurations to the appropriate writer implementation.
"""

from loguru import logger

from .._config import ExportIOConfig, IPFIOConfig, LevelContextIOConfig, RunnerIOConfig

# Export I/O
from .export.osw import OSWReader as ExportOSWReader
from .export.osw import OSWWriter as ExportOSWWriter
from .export.sqmass import SqMassWriter as ExportSqMassWriter
from .export.parquet import (
    ParquetReader as ExportParquetReader,
)
from .export.parquet import (
    ParquetWriter as ExportParquetWriter,
)
from .export.split_parquet import (
    SplitParquetReader as ExportSplitParquetReader,
)
from .export.split_parquet import (
    SplitParquetWriter as ExportSplitParquetWriter,
)

# IPF I/O
from .ipf.osw import OSWReader as IPFOSWReader
from .ipf.osw import OSWWriter as IPFOSWWriter
from .ipf.parquet import ParquetReader as IPFParquetReader
from .ipf.parquet import ParquetWriter as IPFParquetWriter
from .ipf.split_parquet import SplitParquetReader as IPFSplitParquetReader
from .ipf.split_parquet import SplitParquetWriter as IPFSplitParquetWriter

# Levels Context I/O
from .levels_context.osw import OSWReader as LevelContextOSWReader
from .levels_context.osw import OSWWriter as LevelContextOSWWriter
from .levels_context.parquet import ParquetReader as LevelContextParquetReader
from .levels_context.parquet import ParquetWriter as LevelContextParquetWriter
from .levels_context.split_parquet import (
    SplitParquetReader as LevelContextSplitParquetReader,
)
from .levels_context.split_parquet import (
    SplitParquetWriter as LevelContextSplitParquetWriter,
)

# Scoring I/O
from .scoring.osw import OSWReader as ScoringOSWReader
from .scoring.osw import OSWWriter as ScoringOSWWriter
from .scoring.parquet import ParquetReader as ParquetScoringReader
from .scoring.parquet import ParquetWriter as ParquetScoringWriter
from .scoring.split_parquet import SplitParquetReader as SplitParquetScoringReader
from .scoring.split_parquet import SplitParquetWriter as SplitParquetScoringWriter
from .scoring.tsv import TSVReader as ScoringTSVReader
from .scoring.tsv import TSVWriter as ScoringTSVWriter


class ReaderDispatcher:
    """
    Dispatcher class to route I/O configuration to the appropriate reader implementation.

    Based on the `file_type` and optionally the config context, this class instantiates and returns
    the correct reader (e.g., OSWReader, ParquetReader, SplitParquetReader).

    Supported file types:
    - "osw"
    - "parquet"
    - "parquet_split"
    - "parquet_split_multi"
    - "tsv" (Note: only supported for scoring)
    """

    @staticmethod
    def get_reader(config):
        """
        Return the appropriate reader instance based on the config's file_type and context.

        Args:
            config (BaseIOConfig): Configuration object with file_type, level, and context.

        Returns:
            BaseReader: An instance of a subclass of BaseReader suitable for the given input type.

        Raises:
            ValueError: If an unsupported file type is provided.
        """
        if config.file_type == "osw":
            return ReaderDispatcher._get_osw_reader(config)
        elif config.file_type == "parquet":
            logger.warning("Parquet input is experimental. Proceed with caution.")
            return ReaderDispatcher._get_parquet_reader(config)
        elif config.file_type in ("parquet_split", "parquet_split_multi"):
            logger.warning("Split parquet input is experimental. Proceed with caution.")
            return ReaderDispatcher._get_split_parquet_reader(config)
        elif config.file_type == "tsv":
            return ReaderDispatcher._get_tsv_reader(config)
        else:
            raise ValueError(f"Unsupported file type: {config.file_type}")

    @staticmethod
    def _get_osw_reader(config):
        if isinstance(config, RunnerIOConfig):
            return ScoringOSWReader(config)
        elif isinstance(config, IPFIOConfig):
            return IPFOSWReader(config)
        elif isinstance(config, LevelContextIOConfig):
            return LevelContextOSWReader(config)
        elif isinstance(config, ExportIOConfig):
            return ExportOSWReader(config)
        else:
            raise ValueError(f"Unsupported config context: {type(config).__name__}")

    @staticmethod
    def _get_parquet_reader(config):
        if isinstance(config, RunnerIOConfig):
            return ParquetScoringReader(config)
        elif isinstance(config, IPFIOConfig):
            return IPFParquetReader(config)
        elif isinstance(config, LevelContextIOConfig):
            return LevelContextParquetReader(config)
        elif isinstance(config, ExportIOConfig):
            return ExportParquetReader(config)
        else:
            raise ValueError(f"Unsupported config context: {type(config).__name__}")

    @staticmethod
    def _get_split_parquet_reader(config):
        if isinstance(config, RunnerIOConfig):
            return SplitParquetScoringReader(config)
        elif isinstance(config, IPFIOConfig):
            return IPFSplitParquetReader(config)
        elif isinstance(config, LevelContextIOConfig):
            return LevelContextSplitParquetReader(config)
        elif isinstance(config, ExportIOConfig):
            return ExportSplitParquetReader(config)
        else:
            raise ValueError(f"Unsupported config context: {type(config).__name__}")

    @staticmethod
    def _get_tsv_reader(config):
        if isinstance(config, RunnerIOConfig):
            return ScoringTSVReader(config)
        elif isinstance(config, IPFIOConfig):
            raise NotImplementedError("TSV IPFReader not implemented.")
        elif isinstance(config, LevelContextIOConfig):
            raise NotImplementedError("TSV LevelsContextReader not implemented.")
        else:
            raise ValueError(f"Unsupported config context: {type(config).__name__}")


class WriterDispatcher:
    """
    Dispatcher class to route I/O configuration to the appropriate writer implementation.

    Based on the `file_type` and optionally the config context, this class instantiates and returns
    the correct writer (e.g., OSWWriter, ParquetWriter, SplitParquetWriter).

    Supported file types:
    - "osw"
    - "parquet"
    - "parquet_split"
    - "parquet_split_multi"
    - "tsv"
    """

    @staticmethod
    def get_writer(config):
        """
        Return the appropriate writer instance based on the config's file_type and context.

        Args:
            config (BaseIOConfig): Configuration object with file_type, level, and context.

        Returns:
            BaseWriter: An instance of a subclass of BaseWriter suitable for the given output type.

        Raises:
            ValueError: If an unsupported file type is provided.
        """
        if config.file_type == "osw":
            return WriterDispatcher._get_osw_writer(config)
        if config.file_type == "sqmass":
            return WriterDispatcher._get_sqmass_writer(config)
        elif config.file_type == "parquet":
            return WriterDispatcher._get_parquet_writer(config)
        elif config.file_type in ("parquet_split", "parquet_split_multi"):
            return WriterDispatcher._get_split_parquet_writer(config)
        elif config.file_type == "tsv":
            return WriterDispatcher._get_tsv_writer(config)
        else:
            raise ValueError(f"Unsupported file type: {config.file_type}")

    @staticmethod
    def _get_osw_writer(config):
        if isinstance(config, RunnerIOConfig):
            return ScoringOSWWriter(config)
        elif isinstance(config, IPFIOConfig):
            return IPFOSWWriter(config)
        elif isinstance(config, LevelContextIOConfig):
            return LevelContextOSWWriter(config)
        elif isinstance(config, ExportIOConfig):
            return ExportOSWWriter(config)
        else:
            raise ValueError(f"Unsupported config context: {type(config).__name__}")

    @staticmethod
    def _get_sqmass_writer(config):
        if isinstance(config, ExportIOConfig):
            return ExportSqMassWriter(config)
        else:
            raise ValueError(f"Unsupported config context: {type(config).__name__}")

    @staticmethod
    def _get_parquet_writer(config):
        if isinstance(config, RunnerIOConfig):
            return ParquetScoringWriter(config)
        elif isinstance(config, IPFIOConfig):
            return IPFParquetWriter(config)
        elif isinstance(config, LevelContextIOConfig):
            return LevelContextParquetWriter(config)
        elif isinstance(config, ExportIOConfig):
            return ExportParquetWriter(config)
        else:
            raise ValueError(f"Unsupported config context: {type(config).__name__}")

    @staticmethod
    def _get_split_parquet_writer(config):
        if isinstance(config, RunnerIOConfig):
            return SplitParquetScoringWriter(config)
        elif isinstance(config, IPFIOConfig):
            return IPFSplitParquetWriter(config)
        elif isinstance(config, LevelContextIOConfig):
            return LevelContextSplitParquetWriter(config)
        elif isinstance(config, ExportIOConfig):
            return ExportSplitParquetWriter(config)
        else:
            raise ValueError(f"Unsupported config context: {type(config).__name__}")

    @staticmethod
    def _get_tsv_writer(config):
        if isinstance(config, RunnerIOConfig):
            return ScoringTSVWriter(config)
        elif isinstance(config, IPFIOConfig):
            raise NotImplementedError("TSV IPFWriter not implemented.")
        elif isinstance(config, LevelContextIOConfig):
            raise NotImplementedError("TSV LevelContextWriter not implemented.")
        else:
            raise ValueError(f"Unsupported config context: {type(config).__name__}")
