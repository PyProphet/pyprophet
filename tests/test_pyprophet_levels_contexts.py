from __future__ import print_function

import os
import shutil
import sqlite3
import subprocess
import sys
from pathlib import Path

import pandas as pd
import pytest

from pyprophet._config import LevelContextIOConfig
from pyprophet.io.dispatcher import ReaderDispatcher

pd.options.display.expand_frame_repr = False
pd.options.display.precision = 4
pd.options.display.max_columns = None

DATA_FOLDER = Path(__file__).parent / "data"


# ================== SHARED FIXTURES ==================
@pytest.fixture
def temp_folder(tmpdir):
    """Fixture providing temporary folder path"""
    return Path(tmpdir.strpath)


@pytest.fixture
def test_data_osw(temp_folder):
    """Fixture providing OSW test file path"""
    src = DATA_FOLDER / "test_data.osw"
    dst = temp_folder / "test_data.osw"
    shutil.copy(src, dst)
    return dst


@pytest.fixture
def test_data_parquet(temp_folder):
    """Fixture providing Parquet test file path"""
    src = DATA_FOLDER / "test_data.parquet"
    dst = temp_folder / "test_data.parquet"
    shutil.copy(src, dst)
    return dst


@pytest.fixture
def test_data_split_parquet(temp_folder):
    """Fixture providing SplitParquet test folder"""
    src = DATA_FOLDER / "test_data.oswpq"
    dst = temp_folder / "test_data.oswpq"
    shutil.copytree(src, dst)
    return dst


@pytest.fixture(params=["osw", "parquet", "split_parquet"])
def input_strategy(request, test_data_osw, test_data_parquet, test_data_split_parquet):
    """Parametrized fixture for different input strategies"""
    strategies = {
        "osw": {
            "path": test_data_osw,
            "reader": "osw",
            "cmd_prefix": f"--in={test_data_osw}",
        },
        "parquet": {
            "path": test_data_parquet,
            "reader": "parquet",
            "cmd_prefix": f"--in={test_data_parquet}",
        },
        "split_parquet": {
            "path": test_data_split_parquet,
            "reader": "parquet_split",
            "cmd_prefix": f"--in={test_data_split_parquet}",
        },
    }
    return strategies[request.param]


# ================== TEST HELPERS ==================
def run_pyprophet_command(cmd, temp_folder):
    """Helper to run pyprophet commands"""
    try:
        return subprocess.check_output(
            cmd, shell=True, stderr=subprocess.STDOUT, cwd=temp_folder
        ).decode()
    except subprocess.CalledProcessError as error:
        print(f"Command failed: {cmd}\n{error.output.decode()}")
        raise


def validate_peptide_results(regtest, input_path, input_type, context):
    """Validate peptide-level results"""
    # config = LevelContextIOConfig(
    #     infile=input_path,
    #     outfile=input_path,
    #     subsample_ratio=1.0,
    #     level="peptide",
    #     context="levels_context",
    # )
    # config.file_type = input_type
    # config.context_fdr = context
    # reader = ReaderDispatcher.get_reader(config)

    if input_type == "osw":
        sort_cols = [
            "context",
            "run_id",
            "peptide_id",
            "score",
            "pvalue",
            "qvalue",
            "pep",
        ]
        with sqlite3.connect(input_path) as con:
            peptide_data = pd.read_sql_query(
                "SELECT * FROM SCORE_PEPTIDE ORDER BY CONTEXT, RUN_ID, PEPTIDE_ID, SCORE, PVALUE, QVALUE, PEP LIMIT 100",
                con,
            )
            peptide_data.columns = [col.lower() for col in peptide_data.columns]
            print("Peptide Data:", file=regtest)
            print(peptide_data.sort_values(sort_cols).head(100), file=regtest)
    else:
        # For Parquet/SplitParquet, use appropriate reader
        cols = [
            "RUN_ID",
            "FEATURE_ID",
            "PEPTIDE_ID",
            f"SCORE_PEPTIDE_{context.upper().replace('-', '_')}_SCORE",
            f"SCORE_PEPTIDE_{context.upper().replace('-', '_')}_P_VALUE",
            f"SCORE_PEPTIDE_{context.upper().replace('-', '_')}_Q_VALUE",
            f"SCORE_PEPTIDE_{context.upper().replace('-', '_')}_PEP",
        ]

        peptide_data = pd.read_parquet(input_path, columns=cols)

        # Convert IDs to nullable integers (Int64)
        for id_col in ["RUN_ID", "FEATURE_ID"]:
            peptide_data[id_col] = peptide_data[id_col].astype("Int64")

        # Filter NULLs and sort consistently
        peptide_data = peptide_data[peptide_data["RUN_ID"].notnull()]
        peptide_data = peptide_data.sort_values(cols).reset_index(drop=True)

        print("Peptide Data:", file=regtest)
        print(peptide_data.head(100), file=regtest)


def validate_protein_results(regtest, input_path, input_type, context):
    """Validate protein-level results"""
    if input_type == "osw":
        sort_cols = [
            "context",
            "run_id",
            "protein_id",
            "score",
            "pvalue",
            "qvalue",
            "pep",
        ]
        with sqlite3.connect(input_path) as con:
            protein_data = pd.read_sql_query(
                "SELECT * FROM SCORE_PROTEIN ORDER BY CONTEXT, RUN_ID, PROTEIN_ID, SCORE, PVALUE, QVALUE, PEP LIMIT 100",
                con,
            )
            protein_data.columns = [col.lower() for col in protein_data.columns]
            print("Protein Data:", file=regtest)
            print(protein_data.sort_values(sort_cols).head(100), file=regtest)
    else:
        # For Parquet/SplitParquet, use appropriate reader
        cols = [
            "RUN_ID",
            "FEATURE_ID",
            "PROTEIN_ID",
            f"SCORE_PROTEIN_{context.upper().replace('-', '_')}_SCORE",
            f"SCORE_PROTEIN_{context.upper().replace('-', '_')}_P_VALUE",
            f"SCORE_PROTEIN_{context.upper().replace('-', '_')}_Q_VALUE",
            f"SCORE_PROTEIN_{context.upper().replace('-', '_')}_PEP",
        ]

        protein_data = pd.read_parquet(input_path, columns=cols)

        # Convert IDs to nullable integers
        for id_col in ["RUN_ID", "FEATURE_ID"]:
            protein_data[id_col] = protein_data[id_col].astype("Int64")

        protein_data = protein_data[protein_data["RUN_ID"].notnull()]
        protein_data = protein_data.sort_values(cols).reset_index(drop=True)

        print("Protein Data:", file=regtest)
        print(protein_data.head(100), file=regtest)


# ================== TEST CASES ==================
@pytest.mark.parametrize("context", ["run-specific", "experiment-wide", "global"])
def test_peptide_levels(input_strategy, temp_folder, regtest, context):
    """Test peptide-level analysis with different contexts"""
    # Build base command
    cmd = f"pyprophet score {input_strategy['cmd_prefix']} --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # Add peptide command
    cmd += f" && pyprophet infer peptide --pi0_lambda=0.001 0 0 {input_strategy['cmd_prefix']} --context={context}"

    # Execute commands
    run_pyprophet_command(cmd, temp_folder)

    # Validate output
    validate_peptide_results(
        regtest, input_strategy["path"], input_strategy["reader"], context
    )


@pytest.mark.parametrize("context", ["run-specific", "experiment-wide", "global"])
def test_protein_levels(input_strategy, temp_folder, regtest, context):
    """Test protein-level analysis with different contexts"""
    # Build base command
    cmd = f"pyprophet score {input_strategy['cmd_prefix']} --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # Add protein command
    cmd += f" && pyprophet infer protein --pi0_lambda=0 0 0 {input_strategy['cmd_prefix']} --context={context}"

    # Execute commands
    run_pyprophet_command(cmd, temp_folder)

    # Validate output
    validate_protein_results(
        regtest,
        input_strategy["path"],
        input_strategy["reader"],
        context,
    )
