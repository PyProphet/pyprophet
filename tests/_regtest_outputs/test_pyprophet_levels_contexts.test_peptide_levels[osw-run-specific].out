Peptide Data:
         context     pep  peptide_id  pvalue  qvalue               run_id   score
0   run-specific  0.0031        1032  0.0029  0.0033 -8670811102654834151  5.5917
1   run-specific  0.0031         109  0.0029  0.0033 -8670811102654834151  5.5429
2   run-specific  1.0000        1104  0.4692  0.4692 -8670811102654834151 -0.8438
3   run-specific  0.0031        1109  0.0029  0.0033 -8670811102654834151  5.5651
4   run-specific  0.0031        1119  0.0029  0.0033 -8670811102654834151  5.7493
..           ...     ...         ...     ...     ...                  ...     ...
95  run-specific  1.0000         416  0.4692  0.4692 -8670811102654834151 -2.0239
96  run-specific  0.0031         504  0.0029  0.0033 -8670811102654834151  5.9861
97  run-specific  0.0031         539  0.0029  0.0033 -8670811102654834151  5.8118
98  run-specific  1.0000          54  0.4692  0.4692 -8670811102654834151 -0.7028
99  run-specific  0.0031         573  0.0029  0.0033 -8670811102654834151  5.3412

[100 rows x 7 columns]
